opt_einsum/sharing.py

```diff
@@ -4,10 +4,37 @@
 import functools
 import numbers
 from collections import Counter
+import threading
 
 from .parser import alpha_canonicalize, parse_einsum_input
 
-_SHARING_STACK = []
+
+_SHARING_STACK = {}
```

Comment:
Consider making this a `defaultdict(list)` to avoid the `setdefault` complexity below?

Reply:
Yes will do.